File: typescript/src/algorithms/fixed-window-counter.ts
import { RateLimiter, RateLimiterStats } from './rate-limiter.interface';

interface Window {
  count: number;
  startTime: number;
}

export class FixedWindowCounter implements RateLimiter {
  private windows: Map<string, Window> = new Map();
  private limit: number;
  private windowSize: number;

  constructor(limit: number, windowSizeMs: number) {
    this.limit = limit;
    this.windowSize = windowSizeMs;
  }

  async allow(key: string): Promise<boolean> {
    const now = Date.now();
    const currentWindow = Math.floor(now / this.windowSize);
    
    let window = this.windows.get(key);
    
    if (!window || Math.floor(window.startTime / this.windowSize) !== currentWindow) {
      window = {
        count: 0,
        startTime: currentWindow * this.windowSize
      };
      this.windows.set(key, window);
    }

    if (window.count < this.limit) {
      window.count++;
      return true;
    }

    return false;
  }

  async reset(key: string): Promise<void> {
    this.windows.delete(key);
  }

  async getStats(key: string): Promise<RateLimiterStats> {
    const now = Date.now();
    const currentWindow = Math.floor(now / this.windowSize);
    const window = this.windows.get(key);
    
    if (!window || Math.floor(window.startTime / this.windowSize) !== currentWindow) {
      return {
        allowed: true,
        limit: this.limit,
        remaining: this.limit,
        resetAt: new Date((currentWindow + 1) * this.windowSize)
      };
    }

    const remaining = Math.max(0, this.limit - window.count);
    const resetAt = new Date((currentWindow + 1) * this.windowSize);

    return {
      allowed: window.count < this.limit,
      limit: this.limit,
      remaining,
      resetAt
    };
  }
}

File: go/algorithms/fixed_window_counter.go
package algorithms

import (
	"sync"
	"time"
)

type window struct {
	count     int
	startTime time.Time
}

type FixedWindowCounter struct {
	mu         sync.RWMutex
	windows    map[string]*window
	limit      int
	windowSize time.Duration
}

func NewFixedWindowCounter(limit int, windowSize time.Duration) *FixedWindowCounter {
	return &FixedWindowCounter{
		windows:    make(map[string]*window),
		limit:      limit,
		windowSize: windowSize,
	}
}

func (fwc *FixedWindowCounter) Allow(key string) bool {
	fwc.mu.Lock()
	defer fwc.mu.Unlock()

	now := time.Now()
	currentWindow := now.Truncate(fwc.windowSize)

	w, exists := fwc.windows[key]
	if !exists || w.startTime != currentWindow {
		w = &window{
			count:     0,
			startTime: currentWindow,
		}
		fwc.windows[key] = w
	}

	if w.count < fwc.limit {
		w.count++
		return true
	}

	return false
}

func (fwc *FixedWindowCounter) Reset(key string) {
	fwc.mu.Lock()
	defer fwc.mu.Unlock()
	delete(fwc.windows, key)
}

func (fwc *FixedWindowCounter) GetStats(key string) RateLimiterStats {
	fwc.mu.RLock()
	defer fwc.mu.RUnlock()

	now := time.Now()
	currentWindow := now.Truncate(fwc.windowSize)

	w, exists := fwc.windows[key]
	if !exists || w.startTime != currentWindow {
		return RateLimiterStats{
			Allowed:   true,
			Limit:     fwc.limit,
			Remaining: fwc.limit,
			ResetAt:   currentWindow.Add(fwc.windowSize),
		}
	}

	remaining := fwc.limit - w.count
	if remaining < 0 {
		remaining = 0
	}

	return RateLimiterStats{
		Allowed:   w.count < fwc.limit,
		Limit:     fwc.limit,
		Remaining: remaining,
		ResetAt:   currentWindow.Add(fwc.windowSize),
	}
}